### PR TITLE
Add public endpoints for thread safety

### PR DESCRIPTION
**Pull request by choutim**
Add Acquire callback to enable values to synchronize access
Add atomic GetOrAdd function

### CODE REVIEW

**Comment by davidlizeng** [comment]
Shouldn't the imports for the simplelru reference rubrikinc instead of choutim?

File: lru.go [hunk 1]
```diff
@@ -17,10 +17,18 @@ func New(size int) (*Cache, error) {
 	return NewWithEvict(size, nil)
 }
 
-// NewWithEvict constructs a fixed size cache with the given eviction
-// callback.
-func NewWithEvict(size int, onEvicted func(key interface{}, value interface{})) (*Cache, error) {
-	lru, err := simplelru.NewLRU(size, simplelru.EvictCallback(onEvicted))
+// NewWithAcquireAndEvict constructs a fixed size cache with the given eviction
+// and acquire callbacks.
+func NewWithAcquireAndEvict(
```
Comment (tongcharlie) [inline lru.go:22]: swap order of args `onEvicted` and `onAcquire` to match order in function name

File: lru.go [hunk 2]
```diff
@@ -30,13 +38,34 @@ func NewWithEvict(size int, onEvicted func(key interface{}, value interface{}))
 	return c, nil
 }
 
+// NewWithEvict constructs a fixed size cache with the given eviction
+// callback.
+func NewWithEvict(
+	size int,
+	onEvicted func(key interface{}, value interface{}),
+) (*Cache, error) {
+	return NewWithAcquireAndEvict(size, nil, nil)
```
Comment (tongcharlie) [inline lru.go]: should this be passing `onEvicted` to `NewWithAcquireAndEvict`?

File: simplelru/lru.go [hunk 3]
```diff
@@ -22,8 +27,11 @@ type entry struct {
 	value interface{}
 }
 
-// NewLRU constructs an LRU of the given size
-func NewLRU(size int, onEvict EvictCallback) (*LRU, error) {
+func NewLRUWithAcquire(
```
Comment (tongcharlie) [inline simplelru/lru.go]: Rename to `NewLRUWithAcquireAndEvict`

File: simplelru/lru.go [hunk 4]
```diff
@@ -22,8 +27,11 @@ type entry struct {
...
+	size int,
+	onEvict EvictCallback,
```
Comment (tongcharlie) [inline simplelru/lru.go]: Also swap order of args here

File: simplelru/lru.go [hunk 5]
```diff
@@ -32,10 +40,16 @@ func NewLRU(size int, onEvict EvictCallback) (*LRU, error) {
 		evictList: list.New(),
 		items:     make(map[interface{}]*list.Element),
 		onEvict:   onEvict,
+		onAcquire: onAcquire,
 	}
 	return c, nil
 }
 
+// NewLRU constructs an LRU of the given size
+func NewLRU(size int, onEvict EvictCallback) (*LRU, error) {
```
Comment (tongcharlie) [inline simplelru/lru.go]: `NewLRUWithEvict`?

**Review by tongcharlie — APPROVED** [review]

**Pull request merged**